single_kernel_mongo/events/ldap.py

```
@@ -0,0 +1,118 @@
+        self.framework.observe(self.restart_if_ready_event, self._on_restart_if_ready)
+
+        self.framework.observe(
+            self.charm.on["ldap-peers"].relation_changed, self._on_restart_if_ready
+        )
+
+    def _on_ldap_ready(self, event: LdapReadyEvent):
+        """Handles the ops event that indicates that ldap relation is ready."""
+        action = "ldap-ready"
+        try:
+            self.manager.on_ldap_ready(event.relation)
+        except WaitingForLdapDataError as err:
+            self.charm.status_manager.set_and_share_status(BlockedStatus("Waiting for LDAP data."))
+            defer_event_with_info_log(logger, event, action, f"{err}")
+
+        except (DeferrableError, DeferrableFailedHookChecksError) as err:
+            defer_event_with_info_log(logger, event, action, f"{err}")
+        except LDAPSNotEnabledError:
+            self.charm.status_manager.set_and_share_status(
+                BlockedStatus("LDAPS not enabled on LDAP application.")
+            )
+        except NonDeferrableFailedHookChecksError as err:
+            logger.error(f"{err}")
+            self.charm.status_manager.set_and_share_status(BlockedStatus(err.args[0]))
+
+    def _on_ldap_unavailable(self, event: LdapUnavailableEvent):
+        """Handles the ops event that indicates that ldap relation is now unavailable."""
+        self.manager.on_ldap_unavailable()
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent):
+        """Handles the ops event that indicates that ldap-certificates relation is ready."""
+        try:
+            self.manager.on_certificate_available(event.certificate, event.ca, event.chain)
+        except DeferrableFailedHookChecksError as err:
+            defer_event_with_info_log(logger, event, "ldap-cert-ready", f"{err}")
+        except NonDeferrableFailedHookChecksError as err:
+            logger.error(f"{err}")
+            self.charm.status_manager.set_and_share_status(BlockedStatus(err.args[0]))
```

What scenarios does this error occur?

I think in a scenario like this where we cannot retry the event we should: fail noisily and raise

---

This is the case where the role is invalid.
I think it's sufficient to go to blocked with the correct error message.

I don't like raising to the top because it breaks the status display and it's painful to recover.
I'd rather set blocked status with correct information so that in the end it will be easier for the operator to recover: they have integrated the wrong relation, let's remove the relation and everything will be good again, which would be way harder to fix with a raise that would put the charm in error state and make the deletion of the integration painful.

---

Thanks for the explaination. For readability, can we set this to .set_and_share_status(BlockedStatus("Invalid role <...>"))

---

Sure but this will not be sturdy if we have more cases that can raise this.
Let me rephrase this case (and following case) in a better manner.

---

Ah - okay! I see - it is probably a nicer developer interface to have it this way (especially if there are many cases raising)

What if each error had an error code associated with it, and each code was associated with a status (in the future status config file)? Or is that over complex? If so maybe you can add a comment here along the lines of:

```
# status is set as a predefined value in the raised exception and not simply an error raised from an underlying API
```